### PR TITLE
Add flash update form

### PR DESCRIPTION
**Pull request by AdityaKhatri**
## This PR doesn't introduce:
- [ ] typos
- [ ] conflict markers
- [ ] unwanted comments
- [ ] temporary files, auto-generated files or secret keys
- [ ] `console.log` meant for debugging
- [ ] codegen errors

### CODE REVIEW

File: src/views/FlashUpdateForm/schema.ts [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        schema = addCondition(
+            schema,
+            value,
+            ['country_district'] as const,
+            ['country_district'] as const,
+            (val): Pick<FormSchemaFields, 'country_district'> => {
+                const countryIds = val?.country_district
+                    ?.map((country) => country.country).filter(isDefined);
+
+                console.log('here', countryIds);
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/schema.ts]: Let's remove console.log

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/views/FlashUpdateForm/ActionsTab/ActionInput/index.tsx [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+const defaultValue: PartialActionTaken = {
+    client_id: randomString(),
+};
+
+interface Props {
+    value: PartialActionTaken;
+    error: ArrayError<PartialActionTaken> | undefined;
+    onChange: (value: SetValueArg<PartialActionTaken>, index: number) => void;
+    index: number;
+    actionOptions: Action[];
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/ActionsTab/ActionInput/index.tsx]:
```suggestion
    options: Action[];
```

File: src/views/FlashUpdateForm/ActionsTab/ActionInput/index.tsx [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        placeholder,
+    } = props;
+
+    const onFieldChange = useFormObject(index, onChange, defaultValue);
+    const error = (value && value.client_id && errorFromProps)
+        ? getErrorObject(errorFromProps?.[value.client_id])
+        : undefined;
+
+    return (
+        <div className={styles.actionInput}>
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/ActionsTab/ActionInput/index.tsx:55]: Let's add NonFieldError

File: src/views/FlashUpdateForm/ActionsTab/ActionInput/index.tsx [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                error={error?.summary}
+                placeholder={placeholder}
+                disabled={disabled}
+            />
+            <Checklist
+                name="actions"
+                onChange={onFieldChange}
+                options={actionOptions}
+                labelSelector={(d) => d.name ?? ''}
+                keySelector={(d) => d.id}
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/ActionsTab/ActionInput/index.tsx]: Let's not inline functions

File: src/views/FlashUpdateForm/ActionsTab/index.tsx [hunk 5]
```diff
@@ -0,0 +1,142 @@
+import { useMemo } from 'react';
+import {
+    EntriesAsList,
+    Error,
+    useFormArray,
+    getErrorObject,
+} from '@togglecorp/toggle-form';
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/ActionsTab/index.tsx]:
```suggestion
import { useMemo } from 'react';
import {
    type EntriesAsList,
    type Error,
    useFormArray,
    getErrorObject,
} from '@togglecorp/toggle-form';
```

File: src/views/FlashUpdateForm/ActionsTab/index.tsx [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+import Container from '#components/Container';
+import useTranslation from '#hooks/useTranslation';
+import InputSection from '#components/InputSection';
+import { useRequest } from '#utils/restRequest';
+import { type GoApiResponse } from '#utils/restRequest';
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/ActionsTab/index.tsx]: Let's keep this in single import

File: src/views/FlashUpdateForm/ActionsTab/index.tsx [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+        const actionMap = {
+            NTLS: actionList.filter(getFilterOrganization('NTLS')),
+            PNS: actionList.filter(getFilterOrganization('PNS')),
+            FDRN: actionList.filter(getFilterOrganization('FDRN')),
+            GOV: actionList.filter(getFilterOrganization('GOV')),
+        };
+
+        return actionMap;
+    }, [actionsResponse?.results]);
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/ActionsTab/index.tsx:81]: We might have similar logic in field update. Let's see if we can re-use this.

File: src/views/FlashUpdateForm/ActionsTab/index.tsx [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+            FDRN: strings.flashUpdateFormActionTakenByIfrcLabel,
+            GOV: strings.flashUpdateFormActionTakenByGovernmentLabel,
+        },
+        {
+            NTLS: strings.flashUpdateFormActionTakenByNationalSocietyDescription,
+            PNS: strings.flashUpdateFormActionTakenByRcrcDescription,
+            FDRN: strings.flashUpdateFormActionTakenByIfrcDescription,
+            GOV: strings.flashUpdateFormActionTakenByGovernmentDescription,
+        },
+    ] as const), [strings]);
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/ActionsTab/index.tsx]: Let's not use array returns

File: src/views/FlashUpdateForm/ActionsTab/index.tsx [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+    return (
+        <Container
+            heading={strings.flashUpdateFormActionTakenTitle}
+        >
+            {value?.actions_taken?.map((actionTaken, index) => (
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/ActionsTab/index.tsx:118]: Let's add NonFieldErrors here

File: src/views/FlashUpdateForm/ContextTab/CountryProvinceInput/index.tsx [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+            country: newValue,
+            district: [],
+        }, index);
+    }, [
+        value,
+        onChange,
+        index,
+    ]);
+
+    const handleDistrictChange = useCallback((newValue: number[] | undefined) => {
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/ContextTab/CountryProvinceInput/index.tsx:69]: We might not need this function anymore.

File: src/views/FlashUpdateForm/ContextTab/CountryProvinceInput/index.tsx [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+        }, index);
+    }, [
+        value,
+        onChange,
+        index,
+    ]);
+
+    return (
+        <div className={styles.countryDistrictInput}>
+            <CountrySelectInput
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/ContextTab/CountryProvinceInput/index.tsx:84]: Let's add NonFieldErrors

File: src/views/FlashUpdateForm/ContextTab/MultiImageWithCaptionInput/index.tsx [hunk 12]
```diff
@@ -0,0 +1,190 @@
+import { useMemo, useCallback } from 'react';
+import {
+    ArrayError,
+    getErrorObject,
+    useFormArray,
+    SetValueArg,
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/ContextTab/MultiImageWithCaptionInput/index.tsx]:
```suggestion
import {
    type ArrayError,
    getErrorObject,
    useFormArray,
    type SetValueArg,
```

File: src/views/FlashUpdateForm/ContextTab/MultiImageWithCaptionInput/index.tsx [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+                                {
+                                    client_id: String(fileId),
+                                    id: fileId,
+                                },
+                                index,
+                            );
+                        }
+                    },
+                );
+            }
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/ContextTab/MultiImageWithCaptionInput/index.tsx:89]: Why are we calling individual setFieldValue and what happens when oldValue is defined?

File: src/views/FlashUpdateForm/ContextTab/MultiImageWithCaptionInput/index.tsx [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+                fileIdToUrlMap={fileIdToUrlMap}
+                setFileIdToUrlMap={setFileIdToUrlMap}
+                icons={icons}
+                actions={actions}
+                hidePreview
+                disabled={disabled}
+            >
+                {label}
+            </GoMultiFileInput>
+            {value && (
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/ContextTab/MultiImageWithCaptionInput/index.tsx]:
```suggestion
            {value && value.length > 0 && (
```

File: src/views/FlashUpdateForm/ContextTab/MultiImageWithCaptionInput/index.tsx [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+                <div className={styles.fileList}>
+                    {value?.map((fileValue, index) => {
+                        if (isNotDefined(fileValue.id)) {
+                            return null;
+                        }
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/ContextTab/MultiImageWithCaptionInput/index.tsx:147]: Why are we checking for `id`? Do we always upload the data and then attach them? In that case, wouldn't we always have id?

File: src/views/FlashUpdateForm/ContextTab/ReferenceInput/index.tsx [hunk 16]
```diff
@@ -0,0 +1,116 @@
+import React from 'react';
+import { randomString } from '@togglecorp/fujs';
+import {
+    ArrayError,
+    useFormObject,
+    type SetValueArg,
+    getErrorObject,
+} from '@togglecorp/toggle-form';
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/ContextTab/ReferenceInput/index.tsx]:
```suggestion
import {
    type ArrayError,
    useFormObject,
    type SetValueArg,
    getErrorObject,
} from '@togglecorp/toggle-form';
```

File: src/views/FlashUpdateForm/ContextTab/ReferenceInput/index.tsx [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+import useTranslation from '#hooks/useTranslation';
+
+import { PartialReferenceType } from '../../schema';
+
+import i18n from './i18n.json';
+import styles from './styles.module.css';
+
+const defaultFormValues: PartialReferenceType = {
+    client_id: randomString(),
+};
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/ContextTab/ReferenceInput/index.tsx]:
```suggestion
const defaultFormValues: PartialReferenceType = () => ({
    client_id: randomString(),
});
```

File: src/views/FlashUpdateForm/ContextTab/ReferenceInput/index.tsx [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+    } = props;
+    const strings = useTranslation(i18n);
+
+    const onValueChange = useFormObject(index, onChange, defaultFormValues);
+    const error = (value && value?.client_id && errorFromProps)
+        ? getErrorObject(errorFromProps?.[value?.client_id])
+        : undefined;
+
+    return (
+        <div className={styles.reference}>
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/ContextTab/ReferenceInput/index.tsx:56]: Let's add NonFieldError

File: src/views/FlashUpdateForm/schema.ts [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+type FlashUpdateFormFields = Omit<
+    FlashUpdateBody,
+    'country_district' | 'references' | 'map_files' | 'graphics_files' | 'actions_taken'
+> & {
+    'country_district': CountryDistrictType[];
+    'references': ReferenceType[];
+    'map_files': MapType[];
+    'graphics_files': GraphicType[];
+    'actions_taken': ActionType[];
+};
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/schema.ts]: Let's use DeepReplace

File: src/views/FlashUpdateForm/schema.ts [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+                    }),
+                }),
+            },
+        };
+
+        schema = addCondition(
+            schema,
+            value,
+            ['country_district'] as const,
+            ['country_district'] as const,
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/schema.ts]: This condition is very weird.
Also we do not need `const`.

File: src/views/FlashUpdateForm/ContextTab/index.tsx [hunk 21]
```diff
@@ -0,0 +1,298 @@
...
+            .join(' - ');
+
+        const now = new Date();
+        const mm = (now.getMonth() + 1).toString().padStart(2, '0');
+        const yyyy = now.getFullYear().toString();
+
+        const date = `${mm}/${yyyy}`;
+        const selectedHazard = disasterOptions?.find((item) => item.id === value?.hazard_type);
+
+        onValueChange(`${countriesTitles} - ${selectedHazard?.name}  ${date}`, 'title');
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/ContextTab/index.tsx]:
```suggestion
        onValueChange(`${countriesTitles || 'Countries'} - ${selectedHazard?.name || 'Hazard'}  ${date}`, 'title');
```

File: src/views/FlashUpdateForm/ContextTab/index.tsx [hunk 22]
```diff
@@ -0,0 +1,298 @@
...
+        const countriesTitles = value.country_district
+            ?.map((country) => (country.country ? countryTitleMapById[country.country] : undefined))
+            .filter(isDefined)
+            .join(' - ');
+
+        const now = new Date();
+        const mm = (now.getMonth() + 1).toString().padStart(2, '0');
+        const yyyy = now.getFullYear().toString();
+
+        const date = `${mm}/${yyyy}`;
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/ContextTab/index.tsx]: We might already have a util for this

File: src/views/FlashUpdateForm/ContextTab/index.tsx [hunk 23]
```diff
@@ -0,0 +1,298 @@
...
+                    disabled={disabled}
+                />
+            </InputSection>
+            <InputSection
+                title={strings.flashUpdateFormContextReferenceTitle}
+                description={strings.flashUpdateFormContextReferenceDescription}
+                multiRow
+                oneColumn
+            >
+                {value.references?.map((reference, index) => (
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/ContextTab/index.tsx:264]: Let's add NonFieldErrors

File: src/views/FlashUpdateForm/FocalPointsTab/index.tsx [hunk 24]
```diff
@@ -0,0 +1,130 @@
...
+        onValueChange,
+        value,
+        disabled,
+    } = props;
+
+    const error = getErrorObject(formError);
+
+    return (
+        <>
+            <Container>
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/FocalPointsTab/index.tsx:37]: Let's add NonFieldErrors

File: src/views/FlashUpdateForm/index.tsx [hunk 25]
```diff
@@ -1,17 +1,430 @@
...
+
+function getNextStep(currentStep: TabKeys, minSteps: number, maxSteps: number) {
+    const next = bound(tabStepMap[currentStep] + 1, minSteps, maxSteps) as TabNumbers;
+    return tabByStepMap[next];
+}
+
+function getPreviousStep(currentStep: TabKeys, minSteps: number, maxSteps: number) {
+    const prev = bound(tabStepMap[currentStep] - 1, minSteps, maxSteps) as TabNumbers;
+    return tabByStepMap[prev];
+}
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/index.tsx]: Let's use the approach used in FieldReportForm

File: src/views/FlashUpdateForm/index.tsx [hunk 26]
```diff
@@ -1,17 +1,430 @@
...
+    ifrc_phone: true,
+};
+
+type ActionWithoutSummary = Omit<NonNullable<FormType['actions_taken']>[number], 'summary'>;
+const defaultActionsTaken: ActionWithoutSummary[] = [
+    { client_id: randomString(), organization: 'NTLS', actions: [] },
+    { client_id: randomString(), organization: 'PNS', actions: [] },
+    { client_id: randomString(), organization: 'FDRN', actions: [] },
+    { client_id: randomString(), organization: 'GOV', actions: [] },
+];
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/index.tsx]: Can we check if DREF form also uses similar methodology?

File: src/views/FlashUpdateForm/index.tsx [hunk 27]
```diff
@@ -1,17 +1,430 @@
...
+                country_district: response.country_district?.map((country) => ({
+                    ...country,
+                    client_id: String(country.client_id)
+                        ?? String(country.id) ?? randomString(),
+                })),
+                map_files: response.map_files?.map((map) => ({
+                    ...map,
+                    client_id: String(map.client_id)
+                        ?? String(map.id) ?? randomString(),
+                })),
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/index.tsx]: Let's use the injectClientId helper.
Also, we don't need to use the id

File: src/views/FlashUpdateForm/index.tsx [hunk 28]
```diff
@@ -1,17 +1,430 @@
...
+            alert.show(
+                strings.flashUpdateFormRedirectMessage,
+                { variant: 'success' },
+            );
+            navigate(
+                generatePath(
+                    flashUpdateFormEditRoute.absolutePath,
+                    { flashUpdateId: response.id },
+                ),
+            );
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/index.tsx]: Let's add a NOTE to navigate to details page.

File: src/views/FlashUpdateForm/index.tsx [hunk 29]
```diff
@@ -1,17 +1,430 @@
...
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/index.tsx]: Let's add a NOTE to navigate to details page.

File: src/views/FlashUpdateForm/index.tsx [hunk 30]
```diff
@@ -1,17 +1,430 @@
...
+                                setError,
+                                handleSubmit,
+                            )}
+                            disabled={submitPending}
+                            variant="secondary"
+                        >
+                            {/* FIXME: Use translations */}
+                            Submit
+                        </Button>
+                    )}
```
Comment (tnagorra) [inline src/views/FlashUpdateForm/index.tsx]: We should scroll to top when form is submitted. Let's check Field Report form

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/views/FlashUpdateForm/FocalPointsTab/index.tsx (hunk 24 above)
Comment (AdityaKhatri) [reply src/views/FlashUpdateForm/FocalPointsTab/index.tsx:37]: Not required

**Review by tnagorra — APPROVED** [review]

**Pull request merged**